File: localDev.py
import re

def findStr(testStr):
	oFile = open('abc.txt','r')
	result = {}
	findStr = '(?=' + testStr + ')'
	lineNum = 1
	for lineStr in iter(oFile):
		#tempResult = [m.start() for m in re.finditer(findStr, lineStr)]
		## use startswith instead of re generator for easy understanding
		tempResult = [i for i in range(len(lineStr)) if lineStr.startswith(testStr,i)]
		if len(tempResult) > 0:
			result[lineNum] = tempResult
		lineNum += 1
	oFile.close()
	resultStr = ''
	for key in result:
		lineMatches = result[key]
		for matchNum in lineMatches:
			resultStr += str(key) + ',' + str(matchNum) + " "
	print resultStr
def reverse( x):
        """
        :type x: int
        :rtype: int
        """
        maxValue = (1<<31) -1
        res = 0;
        if x > 0:
            strV =  str(x)[::-1]
            res = int(strV)
        else:
            strV = str(x*-1)[::-1]
            res =  int(strV) * -1
        if res > maxValue or res < maxValue * -1:
            return 0
        else:
            return res;
findStr('abc')
#https://leetcode.com/problems/add-two-numbers/description/
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        l1Vals = []
        l2Vals = []
        result = []
        addVal = 0
        l1Node = l1
        l2Node = l2
        while l1Node is not None:
            l1Vals.append(l1Node.val)
            l1Node = l1Node.next
        while l2Node is not None:
            l2Vals.append(l2Node.val)
            l2Node = l2Node.next
        if(len(l1Vals) >= len(l2Vals)):
            largerLength = len(l1Vals)
            largerVals = l1Vals
            smallerLength = len(l2Vals)
            smallerVals = l2Vals
        else:
            largerLength = len(l2Vals)
            largerVals = l2Vals
            smallerLength = len(l1Vals)
            smallerVals = l1Vals
        for i in range(largerLength):
            if(i < smallerLength):
                smallerVal = smallerVals[i]
            else:
                smallerVal = 0
            tempVal = smallerVal + largerVals[i] + addVal
            if(tempVal > 9):
                currentNode = ListNode(tempVal % 10)
                addVal = 1
            else:
                currentNode = ListNode(tempVal)
                addVal = 0
            result.append(currentNode)
        if(addVal != 0):
            currentNode = ListNode(1)
            result.append(currentNode)
        for i in range(0, len(result)):
            result[i].next = result[i-1]
        return result
abc = Solution()
a = ListNode(2)
a.next = ListNode(4)
a.next.next = ListNode(3)
b = ListNode(5)
b.next = ListNode(6)
b.next.next = ListNode(4)
res = abc.addTwoNumbers(a,b)
print res[0].val,res[0].next.val

#https://leetcode.com/problems/generate-parentheses/description/
class Solution2(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        result = []
        self._getParenthesis(n,n,"",result)
        return result
        
    def _getParenthesis(self,leftLNum,leftRNum,inputS,result):
        if leftLNum > leftRNum:
            return
        if leftRNum == 0 and leftLNum == 0:
            result.append(inputS)
            return
        if leftLNum > 0:
            self._getParenthesis(leftLNum -1, leftRNum,inputS+"(",result)
        if leftRNum > 0:
            self._getParenthesis(leftLNum, leftRNum -1,inputS+")",result)

#https://leetcode.com/problems/remove-element/description/
class SolutionRemove(object):
    def removeElement(self, nums, val):
        """
        :type nums: List[int]
        :type val: int
        :rtype: int
        """
        if len(nums) == 0:
            return 0
        i = len(nums) -1
        j = 0
        while j < i:
            if nums[i] == val :
                i = i - 1
            else:
                if nums[j] == val :
                    tempVal = nums[j]
                    nums[j] = nums[i]
                    nums[i] = tempVal
                    i = i -1
                    j = j + 1
                else:
                    j = j + 1
        if(nums[j] == val):
            return j
        else:
            return j + 1

#https://leetcode.com/problems/combination-sum/description/
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        result = []
        candidates = sorted(candidates)
        self._getCombi(result,[],candidates,0,target)
        return result
    def _getCombi(self,result,tempVals,nums,fromIndex,reminder):
        if reminder == 0:
            result.append(tempVals[:])
            return
        elif reminder < 0:
            return
        else:
            for i in range(fromIndex,len(nums)):
                if(reminder - nums[i] < 0):
                    break;
                tempVals.append(nums[i])
                self._getCombi(result,tempVals, nums, i, reminder-nums[i])
                tempVals.pop()

File: localDev.go
func reverse(x int) int {
    signValue := 1;
    if x < 0{
        signValue = -1;
    }
    deltaValue := x;
    resSlice := []int{};
    var res, in int;
    for deltaValue * signValue > 9{
        in = deltaValue % 10;
        deltaValue = deltaValue / 10;
        resSlice = append(resSlice,in);
    }
    resSlice = append(resSlice,deltaValue);
    iLength := len(resSlice);
    for i := 0; i < iLength; i = i+1 {
        res = res * 10 + resSlice[i];
    }
    var max = (1 << 31) - 1;
    if(res *signValue > max){
        return 0;
    }else{
        return res;
    }
}
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    var current ListNode;
    var nextL1, nextL2 *ListNode;
    var tempVal,nextVal,addVal,smallVal,largerLength,smallerLength int;
    var l1Vals,l2Vals,larger,smaller []int;
    var eleArr []ListNode;


	for nextL1 = l1; nextL1 != nil; nextL1 = nextL1.Next{
		l1Vals = append(l1Vals,nextL1.Val);
	}
	for nextL2 = l2; nextL2 != nil; nextL2 = nextL2.Next{
		l2Vals = append(l2Vals,nextL2.Val);
	}

    l1Length := len(l1Vals);
    l2Length := len(l2Vals);

    if(l1Length >= l2Length){
    	largerLength = l1Length;
    	smallerLength = l2Length;
    	larger = l1Vals;
    	smaller = l2Vals;
    }else{
    	largerLength = l2Length;
    	smallerLength = l1Length;
    	larger = l2Vals;
    	smaller = l1Vals;
    }

    for j:=0; j < largerLength; j++{
    	if(j < smallerLength){
    		smallVal = smaller[j];
    	}else{
    		smallVal = 0;
    	}
    	tempVal = larger[j] + smallVal + addVal;

		if(tempVal > 9){
			nextVal = tempVal % 10;
			addVal = 1;
		}else{
			nextVal = tempVal;
			addVal = 0;
		}
		current.Val = nextVal;
		eleArr = append(eleArr,current);
    }
    if(addVal != 0){
    	current.Val = 1;
    	eleArr[len(eleArr) -1].Next = &current;
    	eleArr = append(eleArr,current);
    }
    for i :=0; i < len(eleArr) -1; i = i+1{
    	eleArr[i].Next = &eleArr[i+1];
    }

    return &eleArr[0];
}
/**
*https://leetcode.com/problems/generate-parentheses/description/
*/
func generateParenthesis(n int) []string {
    res := [] string{};
    res = getValidRes(n,n,"",res);
    return res;
}

func getValidRes(leftL int, leftR int, inputStr string, res []string) [] string {
    if leftL > leftR {
        return res;
    }
    if leftL == 0 && leftR == 0 {
        res = append(res,inputStr);
        return res;
    }
    if leftL > 0{
        res = getValidRes(leftL -1 , leftR, inputStr + "(", res);
    }
    if leftR > 0{
        res = getValidRes(leftL , leftR -1, inputStr + ")", res);
    }
    return res;
}
func main(){
  
}

//https://leetcode.com/problems/remove-element/description/
func removeElement(nums []int, val int) int {
    if(len(nums) == 0){
        return 0;
    }
    i := len(nums) - 1
    j := 0
    for j < i {
        if nums[i] == val{
            i --;
        }else{
            if(nums[j] == val){
                tempVal := nums[i]
                nums[i] = nums[j]
                nums[j] = tempVal
                j++
                i--
            }else{
                j++
            }
        }
    }
    if nums[j] == val {
        return j
    }else{
        return j+1;
    }
}

//https://leetcode.com/problems/combination-sum/description/
func combinationSum(candicate []int, target int) [][]int{
    sort.Ints(candicate)
    var result [][]int
    if(len(candicate) == 0 ){
        return result;
    }else{
        tempResult := []int{}
         _getCombination(candicate, &result, tempResult,0,target)
    }
    return result

}

func _getCombination(nums []int, result *[][]int, tempResult []int, fromIndex int, currentTarget int){
    tempV := tempResult[:len(tempResult)]
    if(currentTarget == 0){
        tempR := []int{}
        for _, val := range tempV {
            tempR = append(tempR, val)
        }
        *result = append(*result,tempR)
    }else{
        if currentTarget > 0 {
            for i := fromIndex; i < len(nums); i++ {
                if(currentTarget - nums[i] < 0){
                    break;
                }
                tempV = append(tempV, nums[i])
                _getCombination(nums, result, tempV, i , currentTarget - nums[i])
                tempV = tempV[:len(tempV)-1]
            }
        }else{
            return
        }
    }
}